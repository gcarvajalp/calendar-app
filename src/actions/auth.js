import Swal from "sweetalert2";
import { fetchWithoutToken, fetchWithToken } from "../helpers/fetch";
import { types } from "../types/types";


export const startLogin = (email, password) => {
  return async (dispatch) => {
    
    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const body = await resp.json();

    if (body?.token) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-data', new Date().getTime());
      dispatch(login({
        uid: body.uid,
        name: body.name
      }))
    } else {
      Swal.fire('Error', body.message, 'error');
    }
  }
}

export const startRegister = (name, email, password) => {
  return async (dispatch) => {
    const resp = await fetchWithoutToken('auth/new', { name, email, password }, 'POST');
    const body = await resp.json();

    if (body?.token) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-data', new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }));

    } else {
      Swal.fire('Error', body.message, 'error');
    }
  }
}

export const startChecking = () => {
  return async (dispatch) => {

    const resp = await fetchWithToken('auth/renew');
    const body = await resp.json();

    if (body?.token) {
      localStorage.setItem('token', body.token);
      localStorage.setItem('token-init-data', new Date().getTime());

      dispatch(login({
        uid: body.uid,
        name: body.name
      }))
    } else {
      dispatch(checkingFinish());
      //Swal.fire('Error', body.message, 'error');
    }

  }
}

export const checkingFinish = () => ({
  type: types.authCheckingFinish
});

export const login = (user) => ({
  type: types.authLogin,
  payload: user
});

export const startLogout = () => {
  return async (dispatch) => {

    localStorage.clear();
    dispatch(logout());
  }
}

export const logout = () => ({
  type: types.authLogout
})