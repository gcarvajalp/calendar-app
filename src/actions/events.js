import Swal from "sweetalert2";
import { fetchWithToken } from "../helpers/fetch";
import { prepareEvents } from "../helpers/prepare-events";
import { types } from "../types/types";


export const eventStartAddNew = (event) => {

  return async (dispatch, getState) => {

    const { uid, name } = getState().auth;

    try {

      const resp = await fetchWithToken('event/', event, 'POST');
      const body = await resp.json();

      if (body?.event) {

        event.id = body.event.id;
        event.user = {
          _id: uid,
          name
        }

        dispatch(eventAddNew(event));
      } else {
        Swal.fire('Error', body.message, 'error');
      }

    } catch (error) {
      return Swal.fire('Error', error, 'error');
    }


  }
}

const eventAddNew = (event) => ({
  type: types.eventAddNew,
  payload: event
});

export const eventSetActive = (event) => ({
  type: types.eventSetActive,
  payload: event
});

export const eventClearActiveEvent = () => ({
  type: types.eventClearActiveEvent
});

export const eventStartUpdate = (event) => {
  return async (dispatch) => {
    try {

      const resp = await fetchWithToken(`event/${event.id}`, event, 'PUT');
      const body = await resp.json();

      if (body?.updated) {
        dispatch(eventUpdated(event));
      } else {
        return Swal.fire('Error', body.message, 'error');
      }

    } catch (error) {
      console.error(error);
      Swal.fire('Error', error, 'error');
    }
  }
}

const eventUpdated = (event) => ({
  type: types.eventUpdated,
  payload: event
});

export const eventStartDelete = (event) => {
  return async (dispatch) => {

    try {

      const resp = await fetchWithToken(`event/${event.id}`, {}, 'DELETE');
      const body = await resp.json();

      if (body?.event) {
        dispatch(eventDeleted(event));
      } else {
        return Swal.fire('Error', body.message, 'error');
      }

    } catch (error) {
      return Swal.fire('Error', error, 'error');
    }

  }
}

const eventDeleted = (event) => ({
  type: types.eventDeleted
});

export const eventStartLoading = () => {
  return async (dispatch) => {

    try {

      const resp = await fetchWithToken('event');
      const body = await resp.json();

      const events = prepareEvents(body.events);

      dispatch(eventLoaded(events));

    } catch (error) {
      return Swal.fire('Error', error, 'error');
    }

  }
}

export const eventLoaded = (events) => ({
  type: types.eventLoaded,
  payload: events
});

export const eventLogout = () => ({
  type: types.eventLogout
});