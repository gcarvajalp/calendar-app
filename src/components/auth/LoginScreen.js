import React from 'react';
import './login.css';
import {
  Link,
} from "react-router-dom";


const LoginScreen = () => {
  return (
    <div className="container login-container">
      <div className="row justify-content-center">
        <div className="col-md-6 login-form-1">
          <h3>Ingreso</h3>
          <form>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Correo"
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="ContraseÃ±a"
              />
            </div>
            <div className="form-group">
              <input
                type="submit"
                className="btnSubmit btn-block"
                value="Login"
              />
            </div>

            <Link
              to={`/auth/register`}
              className="link"
            >Create a new account</Link>

          </form>
        </div>
      </div>
    </div>
  )
}

export default LoginScreen;