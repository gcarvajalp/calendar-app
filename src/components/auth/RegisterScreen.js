import React from 'react';
import { useDispatch } from 'react-redux';

import './login.css';
import {
  Link
} from "react-router-dom";
import { useForm } from '../../hooks/useForm';
import { startRegister } from '../../actions/auth';
import Swal from 'sweetalert2';

const RegisterScreen = () => {

  const dispatch = useDispatch();

  const [formValues, handleInputChange] = useForm({
    name: '',
    email: '',
    password: '',
    password2: ''
  });

  const { name, email, password, password2 } = formValues;

  const handleRegisterSubmit = (e) => {
    e.preventDefault();

    if (password !== password2) {
      return Swal.fire('Error', 'contraseñas no coinciden', 'warning');
    }

    dispatch(startRegister(name, email, password));
  }

  return (
    <div className="container login-container">
      <div className="row justify-content-center">

        <div className="col-md-6 login-form-2">
          <h3>Registro</h3>
          <form onSubmit={handleRegisterSubmit}>
            <div className="form-group">
              <input
                type="text"
                className="form-control"
                placeholder="Nombre"
                name="name"
                value={name}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="email"
                className="form-control"
                placeholder="Correo"
                name="email"
                value={email}
                onChange={handleInputChange}
              />
            </div>
            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Contraseña"
                name="password"
                value={password}
                onChange={handleInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="password"
                className="form-control"
                placeholder="Repita la contraseña"
                name="password2"
                value={password2}
                onChange={handleInputChange}
              />
            </div>

            <div className="form-group">
              <input
                type="submit"
                className="btnSubmit btn-block"
                value="Crear cuenta" />
            </div>

            <Link
              to={`auth/register`}
              className="link"
            >
              Already registered?
            </Link>

          </form>
        </div>
      </div>
    </div>
  )
}

export default RegisterScreen;