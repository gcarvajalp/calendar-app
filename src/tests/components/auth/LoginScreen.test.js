import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import LoginScreen from '../../../components/auth/LoginScreen';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startLogin } from '../../../actions/auth';

jest.mock('../../../actions/auth', () => ({
  startLogin: jest.fn()
}));

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {};
const store = mockStore(initState);
store.dispatch = jest.fn();

//store.dispatch
const wrapper = mount(
  <Provider store={store}>
    <MemoryRouter>
      <LoginScreen />
    </MemoryRouter>
  </Provider>
);

describe('<LoginScreen/>', () => {

  test('snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('debe de llamar el dispatch del login', () => {

    wrapper.find('input[name="email"]').simulate('change', {
      target: {
        name: 'email',
        value: 'test@gmail.com'
      }
    });

    wrapper.find('input[name="password"]').simulate('change', {
      target: {
        name: 'password',
        value: '123456'
      }
    });

    wrapper.find('form').prop('onSubmit')({
      preventDefault() { }
    });

    expect(startLogin).toHaveBeenCalledWith('test@gmail.com','123456');

  });

})
