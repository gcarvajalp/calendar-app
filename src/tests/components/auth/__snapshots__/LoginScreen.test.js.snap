// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginScreen/> snapshot 1`] = `
<div
  className="container login-container"
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-md-6 login-form-1"
    >
      <h3>
        Ingreso
      </h3>
      <form
        onSubmit={[Function]}
      >
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="email"
            onChange={[Function]}
            placeholder="Correo"
            type="text"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="form-control"
            name="password"
            onChange={[Function]}
            placeholder="ContraseÃ±a"
            type="password"
            value=""
          />
        </div>
        <div
          className="form-group"
        >
          <input
            className="btnSubmit btn-block"
            type="submit"
            value="Login"
          />
        </div>
        <a
          className="link"
          href="/auth/register"
          onClick={[Function]}
        >
          Create a new account
        </a>
      </form>
    </div>
  </div>
</div>
`;
